services/streammanager/lambda/lambdamanager.go

```diff
@@ -0,0 +1,100 @@
+
+package lambda
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/aws/aws-sdk-go/service/lambda"
+	"github.com/rudderlabs/rudder-server/services/streammanager/common"
+	"github.com/rudderlabs/rudder-server/utils/awsutils"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/tidwall/gjson"
+)
+
+// Config is the config that is required to send data to Lambda
+type Config struct {
+	InvocationType string
+	ClientContext  string
+	Lambda         string
+}
+
+type LambdaProducer struct {
+	client LambdaClient
+}
+
+type LambdaClient interface {
+	Invoke(input *lambda.InvokeInput) (*lambda.InvokeOutput, error)
+}
+
+var pkgLogger logger.LoggerI
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child(lambda.ServiceName)
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destinationConfig map[string]interface{}, o common.Opts) (*LambdaProducer, error) {
+	sessionConfig, err := awsutils.NewSessionConfig(destinationConfig, o.Timeout, lambda.ServiceName)
+	if err != nil {
+		return nil, err
+	}
+	if sessionConfig.Region == "" {
+		return nil, errors.New("could not find region configuration")
+	}
+	awsSession, err := awsutils.CreateSession(sessionConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &LambdaProducer{client: lambda.New(awsSession)}, nil
+}
+
+// Produce creates a producer and send data to Lambda.
+func (producer *LambdaProducer) Produce(jsonData json.RawMessage, _ interface{}) (int, string, string) {
+	client := producer.client
+	if client == nil {
+		return simpleErrorResponse("Could not create client")
+	}
+	data := gjson.GetBytes(jsonData, "payload").String()
+	if data == "" {
+		return simpleErrorResponse("Invalid payload")
+	}
+	destConfig := gjson.GetBytes(jsonData, "destConfig").String()
+	if destConfig == "" {
+		return simpleErrorResponse("Invalid Destination Config")
+	}
+
+	var config Config
+	err := json.Unmarshal([]byte(destConfig), &config)
+	if err != nil {
+		return simpleErrorResponse("error while unmarshalling destination config :: " + err.Error())
+	}
+
+	var invokeInput lambda.InvokeInput
+	invokeInput.SetFunctionName(config.Lambda)
+	invokeInput.SetPayload([]byte(data))
+	invokeInput.SetInvocationType(config.InvocationType)
+	if config.ClientContext != "" {
+		invokeInput.SetClientContext(config.ClientContext)
+	}
+
+	if err = invokeInput.Validate(); err != nil {
+		return simpleErrorResponse("Invalid invokeInput :: " + err.Error())
+	}
+
+	_, err = client.Invoke(&invokeInput)
+	if err != nil {
+		statusCode, respStatus, responseMessage := common.ParseAWSError(err)
+		pkgLogger.Errorf("[Lambda] Invocation error :: %d : %s : %s", statusCode, respStatus, responseMessage)
+		return statusCode, respStatus, responseMessage
+	}
+
+	return 200, "Success", "Event delivered to Lambda :: " + config.Lambda
+}
+
+func simpleErrorResponse(message string) (int, string, string) {
```

Comment:
Rename the function to `lambdaErrorResponse` ?

Reply:
terminal error too, since a status code `400` means no retry :)

Reply:
@atzoum This function is called only when we need to send an error response with 400 status code (no retry).

Reply:
The package name is already `lambda`, so adding it as a prefix won't be of any benefit.

Unrelated to this PR, I believe we need to introduce custom error types with the following interface:

```go
type DestinationError interface {
    Error() string // for returnMessage
    Status() string
    Retriable() bool 
}

// HTTPDestinationError contains additional information for destinations that use HTTP protocol
type HTTPDestinationError interface {
    DestinationError
    HTTPCode() int
}
```

A single error can be returned in all cases, instead of the triple, plus knowing when to retry or not and decoupling from HTTP status codes.